### PR TITLE
Fixed bpython workaround to avoid RuntimeError: dictionary changed size during iteration, also fixed tests

### PR DESCRIPTION
**Pull request by DomWeldon**
## Summary 

* Fixed `bpython` workaround to avoid `RuntimeError: dictionary changed size during iteration`
* Adjusted tests to use more modern PyTest 4+ context manager interface for `pytest.raises()`
* Refactored `TestRealModule.setup_class()` into a function-level pytest fixture, using the built-in `tmp_path` fixture.

## `bpython` Error Summary

Attempting to import any package which imports `apipkg` in the [popular REPL bpython](https://bpython-interpreter.org/) will result in an error like the below (trying to import pytest in a poetry virtualenv). 

    >>> import pytest
    Traceback (most recent call last):
      File "<input>", line 1, in <module>
        import pytest
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/pytest.py", line 6, in <module>
        from _pytest.assertion import register_assert_rewrite
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/_pytest/assertion/__init__.py", line 6, in <module>
        from _pytest.assertion import rewrite
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/_pytest/assertion/rewrite.py", line 20, in <module>
        from _pytest.assertion import util
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/_pytest/assertion/util.py", line 5, in <module>
        import _pytest._code
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/_pytest/_code/__init__.py", line 2, in <module>
        from .code import Code  # noqa
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/_pytest/_code/code.py", line 12, in <module>
        import py
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/py/__init__.py", line 153, in <module>
        'Syslog'             : '._log.log:Syslog',
      File "/home/domweldon/.cache/pypoetry/virtualenvs/polymer-versioning-py3.7/lib/python3.7/site-packages/py/_vendored_packages/apipkg.py", line 63, in initpkg
        for module in sys.modules.values():
    RuntimeError: dictionary changed size during iteration

With thanks to [William Schwartz's comment](https://bugs.python.org/issue29620), I believe that this error occurs because the lines below. 

https://github.com/decisionLabLtd/apipkg/blob/4b4b20b8bef95b78cc95f2c77715a20e64e871cb/src/apipkg/__init__.py#L64-L72

Calling `__dict__` on all instances of `ApiModule`, results in a call to `__getattr__()` / `__makeattr()`, which under certain conditions uses `importobj()` to import new modules into the process. 

https://github.com/decisionLabLtd/apipkg/blob/master/src/apipkg/__init__.py#L142-L165 

This adds new members to `sys.modules`, which will cause the error above to be raised.

This fix prevents the error being thrown by first creating a list of modules which need to be eagerly loaded and then iterating through the separate list. It is not anticipated that any new modules loaded during the eager loading will need to themselves be initialized to prevent monkeypatching errors (this assumption matches the original implementation).

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
yikes

**Comment by RonnyPfannschmidt** [comment]
i'll have to take a look at the failures in a bit before i can merge/release, cant promise to sort it this weekend either

**Comment by DomWeldon** [comment]
Crikey - sorry I missed this too. I'll have a look at the build failure now.

**Comment by RonnyPfannschmidt** [comment]
@DomWeldon gentle ping - do you think you can fix it, it will be a while longer before i have the opportunity to do so myself

**Comment by DomWeldon** [comment]
I'll look into this ASAP, probably later this week!

--------------
*Dom Weldon *

M: (+44) 7903 683 124
E: dom.weldon@gmail.com
Twitter: @DomWeldon

**Comment by swt2c** [comment]
Ping?  It would be great to get a release with these fixes.  :-)

File: src/apipkg/__init__.py [hunk 1]
```diff
@@ -62,9 +62,14 @@ def initpkg(pkgname, exportdefs, attr=None, eager=False):
     sys.modules[pkgname] = mod
     # eagerload in bypthon to avoid their monkeypatching breaking packages
     if 'bpython' in sys.modules or eager:
-        for module in list(sys.modules.values()):
```
Comment (RonnyPfannschmidt) [inline src/apipkg/__init__.py:65]: i just took note that the list is actually already there

**Comment by RonnyPfannschmidt** [comment]
closing as already solved on main

will handle re-vendoring into py after the next release (2.0)

**Pull request closed without merging**